autoextract/aio/errors.py

```diff
@@ -15,3 +22,64 @@ def __str__(self):
         return f"ApiError: {self.status}, message={self.message}, " \
                f"headers={self.headers}, body={self.response_content}"
 
+
+class QueryError(Exception):
+    """ Exception which is raised when a Query-level error is returned.
+    https://doc.scrapinghub.com/autoextract.html#query-level
+    """
+
+    DOMAIN_OCCUPIED_REGEX = re.compile(
+        r".*domain (.+) is occupied, please retry in (.+) seconds.*",
+        re.IGNORECASE
+    )
+
+    RETRIABLE_QUERY_ERROR_MESSAGES = {
+        msg.lower().strip()
+        for msg in [
+            "query timed out",
+            "Downloader error: No response (network5)",
+            "Downloader error: http50",
+            "Downloader error: GlobalTimeoutError",
+            "Proxy error: banned",
+            "Proxy error: internal_error",
+            "Proxy error: nxdomain",
+            "Proxy error: timeout",
+            "Proxy error: ssl_tunnel_error",
+            "Proxy error: msgtimeout",
+            "Proxy error: econnrefused",
+            # Retry errors for AutoExtract API dev server
+            "Error making splash request: ServerDisconnectedError",
+            "Error making splash request: ClientOSError: [Errno 32] Broken pipe",
+        ]
+    }
+
+    def __init__(self, query: dict, message: str):
+        self.query = query
+        self.message = message
+
+    def __str__(self):
+        return f"QueryError: query={self.query}, message={self.message}"
+
+    @property
+    def retriable(self) -> bool:
+        if self.domain_occupied:
+            return True
+
+        return self.message.lower() in self.RETRIABLE_QUERY_ERROR_MESSAGES
+
+    @property
+    def domain_occupied(self) -> Optional[str]:
+        match = self.DOMAIN_OCCUPIED_REGEX.match(self.message)
+        return match.group(1) if match else None
+
+    @property
+    def retry_seconds(self) -> Optional[float]:
```

Comment:
Suggestion. What do you think about creating a class `DomainOccupied`:
* with two members:
  * domain
  * retry_seconds
* That also contains the REGEX and the logic for the match. 
* With a class method to build it, e.g. `from_query_error` that returns None if it is not a domain occupied method. 

And finally, have a single property in `QuerError` `domain_occupied` with return type `Optional[DomainOccupied]`

Reply:
I agree with @ivanprado , but I would only consider it if the `DomainOccupied` logic is used in other places, otherwise, I guess it is fine to keep it here.

Reply:
I've tried to move the Domain Occupied logic to a dedicated place. This made the QueryError exception class a little bit more concise. I've kept the `retry_seconds` property as a convenience in order to keep retry and client logic more simple (e.g. `_get_exception_with_longest_timeout`). Let me know what you think.

Reply:
I see that this change implies more lines added that those removed, so not sure anymore. Anyway, I think that also `retry_seconds` should be removed from RequestError, because having it there suggest that every `RequestError` can have some retry seconds, which is not the case. 

But anyway, this is very minor, just a matter of taste as well. So I'm happy with original version or with this one if you want to keep them.

Reply:
> I see that this change implies more lines added that those removed

Indeed but at least things are a little bit more separated now.

> Anyway, I think that also retry_seconds should be removed from RequestError, because having it there suggest that every RequestError can have some retry seconds, which is not the case.

Not really, it has the same signature as before my refactor. Look that our return type is an **optional float**.

I like `QueryError` having this `retry_seconds` property because it could make it easier to add custom retry seconds depending on the Query-level error type, instead of having to depend on the `retry.py` file.

At the end of the day, I don't have a preference between these two versions neither. Let's wait for @ejulio's input here otherwise, we could resolve this conversation the way it is now.

Reply:
No strong opinion on this one.
If there is no time to retry because it is not retriable, the the Optional makes sense.
If it is because we can run right away, I would simplify and return as 0 instead of `None`, as 0 means no wait time required, while `None` gives the idea of `this is not retriable`

Reply:
Done.